src/lib/fs/conffile.c

```diff
@@ -50,62 +50,97 @@ config_get_lines_include(const char *string, config_line_t **result,
- * only that file will be added to <b>file_list</b>. If it is a directory,
+/** Returns a list of configuration files present on paths that match
+ * <b>pattern</b>. The pattern is expanded and then all the paths are
+ * processed. A path can be a file or a directory. If it is a file, that file
+ * will be added to the list to be returned. If it is a directory,
  * all paths for files on that directory root (no recursion) except for files
- * whose name starts with a dot will be added to <b>file_list</b>.
+ * whose name starts with a dot will be added to the list to be returned.
  * <b>opened_files</b> will have a list of files opened by this function
- * if provided. Return 0 on success, -1 on failure. Ignores empty files.
+ * if provided. Return NULL on failure. Ignores empty files.
  */
 static smartlist_t *
-config_get_file_list(const char *path, smartlist_t *opened_files)
+config_get_file_list(const char *pattern, smartlist_t *opened_files)
 {
-  smartlist_t *file_list = smartlist_new();
+  smartlist_t *matches = tor_glob(pattern);
+  if (!matches) {
+    return NULL;
+  }
 
-  if (opened_files) {
-    smartlist_add_strdup(opened_files, path);
+  // if it is not a glob, return error when the path is missing
+  if (!has_glob(pattern) && smartlist_len(matches) == 0) {
+    smartlist_free(matches);
+    return NULL;
   }
 
-  file_status_t file_type = file_status(path);
-  if (file_type == FN_FILE) {
-    smartlist_add_strdup(file_list, path);
-    return file_list;
-  } else if (file_type == FN_DIR) {
-    smartlist_t *all_files = tor_listdir(path);
-    if (!all_files) {
-      smartlist_free(file_list);
+  if (opened_files) {
+    smartlist_t *glob_opened = get_glob_opened_files(pattern);
+    if (!glob_opened) {
+      SMARTLIST_FOREACH(matches, char *, f, tor_free(f));
+      smartlist_free(matches);
       return NULL;
     }
-    smartlist_sort_strings(all_files);
-    SMARTLIST_FOREACH_BEGIN(all_files, char *, f) {
-      if (f[0] == '.') {
-        tor_free(f);
-        continue;
-      }
+    smartlist_add_all(opened_files, glob_opened);
+    smartlist_free(glob_opened);
+  }
 
-      char *fullname;
-      tor_asprintf(&fullname, "%s"PATH_SEPARATOR"%s", path, f);
-      tor_free(f);
+  smartlist_sort_strings(matches);
+  int error_found = 0;
+  smartlist_t *file_list = smartlist_new();
+  SMARTLIST_FOREACH_BEGIN(matches, char *, path) {
+    if (opened_files) {
+      smartlist_add_strdup(opened_files, path);
+    }
 
-      if (opened_files) {
-        smartlist_add_strdup(opened_files, fullname);
+    file_status_t file_type = file_status(path);
+    if (file_type == FN_FILE) {
+      smartlist_add_strdup(file_list, path);
+    } else if (file_type == FN_DIR) {
+      smartlist_t *all_files = tor_listdir(path);
+      if (!all_files) {
+        error_found = 1;
+        break;
       }
+      smartlist_sort_strings(all_files);
+      SMARTLIST_FOREACH_BEGIN(all_files, char *, f) {
+        if (f[0] == '.') {
```

Comment:
This check doesn't do what the function says it does: if "f" is a path, "f[0]" is the first character of the first directory in the path. But the function documentation says "except for files whose name starts with a dot".

The unit tests are incomplete or incorrect, because they should have caught this case.

I'm also not sure that this design is consistent and easy to understand.
Here's a consistent set of rules that matches typical shell behaviour:
1. dots that are explicitly included in the file or directory path should be included
2. If a glob expands to the start of a path component, and that path component starts with a dot, then that path component (and all files under it) should be excluded

Reply:
This code is run when expanding a directory and f is always a filename.
The wildcard expansion works as you described.
The tests were indeed incomplete, I've added a test for hidden files on 9fd5ccb.